### PR TITLE
End time after start time

### PR DESCRIPTION
**Pull request by joycelynng**
New end-time-after-start-time branch, to replace previous implement-end-time-after-start-time branch.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -72,7 +76,25 @@ public CommandResult execute() {
             if(newName.isPresent()) {
         	    postEdit.setName(newName.get());
             }
-        	
+            
+            if(newStartDateTime.isPresent() && !newEndDateTime.isPresent()){
+            	LocalDateTime startDateTime = newStartDateTime.get();
+            	LocalDateTime endDateTime = taskToEdit.getEndDate().get();
+            	
+            	if(startDateTime.isAfter(endDateTime)){
```
Comment (kabir-plod) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Suggest `!endDateTime.isAfter(startDateTime)` instead to cover the (invalid) case where both are the same

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 2]
```diff
@@ -252,6 +259,40 @@ private Command prepareAddEvent(String arguments) {
 			}
 		}
 
+		ArrayList<Matcher> diffDayMatchers = new ArrayList<>();
```
Comment (kabir-plod) [inline src/main/java/seedu/address/logic/parser/Parser.java]: This part of the parser does not exist in master. Suggest copying the parser source over from master and then inserting changes.

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 3]
```diff
@@ -499,7 +540,12 @@ private Command prepareEdit(String arguments) {
 			}
 		}
 		
-
+		if(startDateTime!=null && endDateTime!=null){
```
Comment (kabir-plod) [inline src/main/java/seedu/address/logic/parser/Parser.java]: Can be simplified into one if statement `if (start != null && end != null && !end.isAfter(start))`

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 4]
```diff
@@ -71,7 +77,25 @@ public CommandResult execute() {
             if(newName.isPresent()) {
         	    postEdit.setName(newName.get());
             }
-        	
+            
+            if(newStartDateTime.isPresent() && !newEndDateTime.isPresent()){
+            	LocalDateTime startDateTime = newStartDateTime.get();
+            	LocalDateTime endDateTime = taskToEdit.getEndDate().get();
```
Comment (kelvinnpc) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: taskToEdit may not have endDate if it is a someday task. getEndDate() will return Optional.empty() in this case and calling the .get() function will give error. Suggest doing a check for getEndDate isPresent before using the .get()

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 5]
```diff
@@ -71,7 +77,25 @@ public CommandResult execute() {
...
+            	
+            	if(startDateTime.isAfter(endDateTime)){
+            		return new CommandResult(MESSAGE_START_DATE_TIME_AFTER_END_DATE_TIME);
+            	}            	
+            }
+            
+            if(!newStartDateTime.isPresent() && newEndDateTime.isPresent()){
+            	LocalDateTime startDateTime = taskToEdit.getStartDate().get();
```
Comment (kelvinnpc) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: same problem as mentioned above. taskToEdit may not have a startDate

**Comment by kabir-plod** [comment]
parseCommand_addEventValidOrder_addCommandReturned failing. Please correct and add new test case to confirm intended behaviour.

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 6]
```diff
@@ -500,7 +511,16 @@ private Command prepareEdit(String arguments) {
 			}
 		}
 		
-
+		if(startDateTime!=null && endDateTime!=null && !endDateTime.isAfter(startDateTime)){
+			if(startDateTime.isAfter(endDateTime)){
+				return new IncorrectCommand(EditCommand.MESSAGE_START_DATE_TIME_AFTER_END_DATE_TIME);
```
Comment (kabir-plod) [inline src/main/java/seedu/address/logic/parser/Parser.java]: Same as above

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 7]
```diff
@@ -240,6 +240,17 @@ private Command prepareAddEvent(String arguments) {
...
 					startDateTime = DateParser.parse(date + " " + startTime);
 					endDateTime = DateParser.parse(date + " " + endTime);
+					
+					if(!endDateTime.isAfter(startDateTime)){
+						if(startDateTime.isAfter(endDateTime)){
+							return new IncorrectCommand(AddCommand.MESSAGE_START_DATE_TIME_AFTER_END_DATE_TIME);
+						}
+						
+						else{
+							return new IncorrectCommand(AddCommand.MESSAGE_START_DATE_TIME_EQUALS_END_DATE_TIME);
```
Comment (kabir-plod) [inline src/main/java/seedu/address/logic/parser/Parser.java]: Nested if/else can be removed - a less precise message will suffice in return for cleaner code

**Pull request merged**